### PR TITLE
Rewrite dist script

### LINKED ISSUE #251: Incorrect sha256sum checksum file
Current oauth2_proxy [sha256sum checksum file](https://github.com/pusher/oauth2_proxy/releases/download/v4.0.0/oauth2_proxy-v4.0.0.linux-amd64-sha256sum.txt) contains path that differs from the structure of the [release tarball](https://github.com/pusher/oauth2_proxy/releases/download/v4.0.0/oauth2_proxy-v4.0.0.linux-amd64.go1.12.1.tar.gz). This results in the checksum verification failure.

## Expected Behavior

1. Following the steps in the [Installation documentatio](https://pusher.github.io/oauth2_proxy/installation#installation) should result in successful binary checksum verification.

## Current Behavior

Checksum verification fails:

```
$ sha256sum -c oauth2_proxy-v4.0.0.linux-amd64-sha256sum.txt 
sha256sum: can't open 'release/oauth2_proxy-v4.0.0.linux-amd64.go1.12.1/oauth2_proxy': No such file or directory
release/oauth2_proxy-v4.0.0.linux-amd64.go1.12.1/oauth2_proxy: FAILED
sha256sum: WARNING: 1 of 1 computed checksums did NOT match
```

The structure of the release tarball:
```
$ tar xvf oauth2_proxy-v4.0.0.linux-amd64.go1.12.1.tar.gz 
oauth2_proxy-v4.0.0.linux-amd64.go1.12.1/
oauth2_proxy-v4.0.0.linux-amd64.go1.12.1/oauth2_proxy
```

## Possible Solution

Create the checksum file that match the directory structure of the released tarball:
* `oauth2_proxy-v4.0.0.linux-amd64.go1.12.1/oauth2_proxy` instead of  `release/oauth2_proxy-v4.0.0.linux-amd64.go1.12.1/oauth2_proxy`

## Steps to Reproduce (for bugs)
1. Follow the steps in the [Installation documentation](https://pusher.github.io/oauth2_proxy/installation#installation) for downloading latest v4.0.0 release
2. sha256sum verification fails

## Your Environment

- Version used: [v4.0.0](https://github.com/pusher/oauth2_proxy/releases/tag/v4.0.0)
- Tested with linuxamd64 artifacts: 
  * https://github.com/pusher/oauth2_proxy/releases/download/v4.0.0/oauth2_proxy-v4.0.0.linux-amd64.go1.12.1.tar.gz
  * https://github.com/pusher/oauth2_proxy/releases/download/v4.0.0/oauth2_proxy-v4.0.0.linux-amd64-sha256sum.txt

### PR DESCRIPTION
**Pull request by syscll**
## Description

Fixes #251 

## Motivation and Context

There was a small bug in the `make release` command that did not properly generate `SHA256` hashes. As the original `dist.sh` script was very outdated, I decided to merge the two.

## How Has This Been Tested?

Ran `dist.sh` on my machine and verified that all binaries are built and all `sha256sum` hashes are correct:

```
$ BINARY=oauth2_proxy VERSION=v4.0.0 ./dist.sh
oauth2_proxy-v4.0.0.darwin-amd64.go1.13.3/
oauth2_proxy-v4.0.0.darwin-amd64.go1.13.3/oauth2_proxy
oauth2_proxy-v4.0.0.linux-amd64.go1.13.3/
oauth2_proxy-v4.0.0.linux-amd64.go1.13.3/oauth2_proxy
oauth2_proxy-v4.0.0.linux-arm64.go1.13.3/
oauth2_proxy-v4.0.0.linux-arm64.go1.13.3/oauth2_proxy
oauth2_proxy-v4.0.0.linux-armv6.go1.13.3/
oauth2_proxy-v4.0.0.linux-armv6.go1.13.3/oauth2_proxy
oauth2_proxy-v4.0.0.windows-amd64.go1.13.3/
oauth2_proxy-v4.0.0.windows-amd64.go1.13.3/oauth2_proxy
$ cd release
$ sha256sum -c oauth2_proxy-v4.0.0.darwin-amd64-sha256sum.txt 2>&1
oauth2_proxy-v4.0.0.darwin-amd64.go1.13.3/oauth2_proxy: OK
$ sha256sum -c oauth2_proxy-v4.0.0.linux-amd64-sha256sum.txt 2>&1
oauth2_proxy-v4.0.0.linux-amd64.go1.13.3/oauth2_proxy: OK
$ sha256sum -c oauth2_proxy-v4.0.0.linux-arm64-sha256sum.txt 2>&1
oauth2_proxy-v4.0.0.linux-arm64.go1.13.3/oauth2_proxy: OK
$ sha256sum -c oauth2_proxy-v4.0.0.linux-armv6-sha256sum.txt 2>&1
oauth2_proxy-v4.0.0.linux-armv6.go1.13.3/oauth2_proxy: OK
$ sha256sum -c oauth2_proxy-v4.0.0.windows-amd64-sha256sum.txt 2>&1
oauth2_proxy-v4.0.0.windows-amd64.go1.13.3/oauth2_proxy: OK
```

@JoelSpeed @steakunderscore please run this script on you local machines in order to validate as part of the review.

## Checklist:

- [x] My change requires a change to the documentation or CHANGELOG.
- [x] I have updated the documentation/CHANGELOG accordingly.
- [x] I have created a feature (non-master) branch for my PR.

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -61,29 +61,4 @@ test: lint
...
-	shasum -a 256 release/$(BINARY)-$(VERSION).linux-arm64.$(GO_VERSION)/$(BINARY) > release/$(BINARY)-$(VERSION).linux-arm64-sha256sum.txt
-	shasum -a 256 release/$(BINARY)-$(VERSION).linux-armv6.$(GO_VERSION)/$(BINARY) > release/$(BINARY)-$(VERSION).linux-armv6-sha256sum.txt
-	shasum -a 256 release/$(BINARY)-$(VERSION).windows-amd64.$(GO_VERSION)/$(BINARY) > release/$(BINARY)-$(VERSION).windows-amd64-sha256sum.txt
-	tar -C release -czvf release/$(BINARY)-$(VERSION).darwin-amd64.$(GO_VERSION).tar.gz $(BINARY)-$(VERSION).darwin-amd64.$(GO_VERSION)
-	tar -C release -czvf release/$(BINARY)-$(VERSION).linux-amd64.$(GO_VERSION).tar.gz $(BINARY)-$(VERSION).linux-amd64.$(GO_VERSION)
-	tar -C release -czvf release/$(BINARY)-$(VERSION).linux-arm64.$(GO_VERSION).tar.gz $(BINARY)-$(VERSION).linux-arm64.$(GO_VERSION)
-	tar -C release -czvf release/$(BINARY)-$(VERSION).linux-armv6.$(GO_VERSION).tar.gz $(BINARY)-$(VERSION).linux-armv6.$(GO_VERSION)
-	tar -C release -czvf release/$(BINARY)-$(VERSION).windows-amd64.$(GO_VERSION).tar.gz $(BINARY)-$(VERSION).windows-amd64.$(GO_VERSION)
+	BINARY=$(BINARY) VERSION=${VERSION} ./dist.sh
```
Comment (JoelSpeed) [inline Makefile]: Why are `$(BINARY)` and `${VERSION}` using different types of brackets? Is there a syntactic difference? 🤔

File: dist.sh [hunk 2]
```diff
@@ -1,45 +1,46 @@
...
+
+mkdir release
+
+# Create architecture specific release dirs
+for ARCH in "${ARCHS[@]}"; do
+	mkdir release/${BINARY}-${VERSION}.${ARCH}.${GO_VERSION}
+done
+
+# Create architecture specific binaries
+GO111MODULE=on GOOS=darwin GOARCH=amd64 go build -ldflags="-X main.VERSION=${VERSION}" \
```
Comment (JoelSpeed) [inline dist.sh]: Would this be easier to reason about if it were a loop approx like

```
for ARCH in "${ARCHS[@]}"; do
	GO_ARCH=$(echo $ARCH | awk -F- '{print $2}')
        GO_OS=$(echo $ARCH | awk -F- '{print $1}')
        GO111MODULE=on GOOS=${GO_OS} GOARCH=${GO_ARCH} go build -ldflags="-X main.VERSION=${VERSION}" \
	-o release/${BINARY}-${VERSION}.linux-amd64.${GO_VERSION}/${BINARY} github.com/pusher/oauth2_proxy
done
```

File: dist.sh [hunk 3]
```diff
@@ -1,45 +1,46 @@
...
+	echo "Go version must be >= go1.13"
+	exit 1
 fi
-touch $checksum_file
-for checksum in "${sha256sum[@]}"; do
-    echo "$checksum" >> $checksum_file
+
+ARCHS=(darwin-amd64 linux-amd64 linux-arm64 linux-armv6 windows-amd64)
+
+mkdir release
```
Comment (JoelSpeed) [inline dist.sh]: Should this be `mkdir -p release` so that it can continue if the folder already exists? 🤔

File: Makefile (hunk 1 above)
Comment (syscll) [reply Makefile]: I think this is just a typo, will fix.

**Comment by syscll** [comment]
@JoelSpeed PTAL

File: dist.sh [hunk 4]
```diff
@@ -1,45 +1,46 @@
...
+
+ARCHS=(darwin-amd64 linux-amd64 linux-arm64 linux-armv6 windows-amd64)
+
+mkdir -p release
+
+# Create architecture specific release dirs
+for ARCH in "${ARCHS[@]}"; do
+	mkdir -p release/${BINARY}-${VERSION}.${ARCH}.${GO_VERSION}
+
+    GO_OS=$(echo $ARCH | awk -F- '{print $1}')
```
Comment (JoelSpeed) [inline dist.sh]: Nit: Indentation is off

**Pull request merged**